File: 2025/03/solution1.lua
#!/usr/bin/env lua

-- Read input
local file = io.open('input1.txt', 'r')
local input = file:read('*all'):match('^%s*(.-)%s*$')
file:close()

-- Parse banks (each line is a bank)
local banks = {}
for line in input:gmatch('[^\r\n]+') do
  table.insert(banks, line:match('^%s*(.-)%s*$'))
end

-- Find maximum joltage for a single bank
local function max_joltage(bank)
  local digits = {}
  for i = 1, #bank do
    table.insert(digits, tonumber(bank:sub(i, i)))
  end
  local max_jolt = 0

  -- Try all pairs of positions (i, j) where i < j
  for i = 1, #digits do
    for j = i + 1, #digits do
      local jolt = digits[i] * 10 + digits[j]
      if jolt > max_jolt then
        max_jolt = jolt
      end
    end
  end

  return max_jolt
end

-- Find maximum joltage for each bank and sum them
local total_joltage = 0
for _, bank in ipairs(banks) do
  total_joltage = total_joltage + max_joltage(bank)
end

print("Total output joltage: " .. total_joltage)

File: 2025/03/solution1.php
#!/usr/bin/env php
<?php

// Read input
$input = trim(file_get_contents('input1.txt'));

// Parse banks (each line is a bank)
$banks = array_filter(array_map('trim', explode("\n", $input)));

// Find maximum joltage for a single bank
function max_joltage($bank) {
    $digits = array_map('intval', str_split($bank));
    $max_jolt = 0;

    // Try all pairs of positions (i, j) where i < j
    for ($i = 0; $i < count($digits); $i++) {
        for ($j = $i + 1; $j < count($digits); $j++) {
            $jolt = $digits[$i] * 10 + $digits[$j];
            $max_jolt = max($max_jolt, $jolt);
        }
    }

    return $max_jolt;
}

// Find maximum joltage for each bank and sum them
$total_joltage = 0;
foreach ($banks as $bank) {
    $total_joltage += max_joltage($bank);
}

echo "Total output joltage: " . $total_joltage . "\n";
